pydm/data_plugins/calc_plugin.py

```diff
@@ -0,0 +1,204 @@
+import threading
+
+from pydm.data_plugins.plugin import PyDMPlugin, PyDMConnection
+from qtpy.QtCore import Slot, QThread, Signal, Qt
+from qtpy.QtWidgets import QApplication
+
+import pydm
+
+logger = logging.getLogger(__name__)
+
+
+class CalcThread(QThread):
+    new_data_signal = Signal(dict)
+
+    def __init__(self, config, *args, **kwargs):
+        QThread.__init__(self, *args, **kwargs)
+        self.app = QApplication.instance()
+        self.app.aboutToQuit.connect(self.requestInterruption)
+
+        self.config = config
+
+        self._calculate = threading.Event()
+        self._channels = []
+        self._value = None
+        self._values = None
+        self._connections = None
+        self._expression = self.config.get('expr', '')
+
+        for ch_idx, ch in enumerate(self.config.get('channels', [])):
+            conn_cb = functools.partial(self.callback_conn, ch_idx)
+            value_cb = functools.partial(self.callback_value, ch_idx)
+            c = pydm.PyDMChannel(ch, connection_slot=conn_cb,
+                                 value_slot=value_cb)
+            self._channels.append(c)
+
+    @property
+    def connected(self):
+        return all(self._connections)
+
+    def _connect(self):
+        self._values = [None]*len(self._channels)
+        self._connections = [False]*len(self._channels)
+        for ch in self._channels:
+            ch.connect()
+
+    def _disconnect(self):
+        for ch in self._channels:
+            ch.disconnect()
+
+    def _send_update(self, conn, value):
+        self.new_data_signal.emit({"connection": conn,
+                                   "value": value})
+
+    def run(self):
+        self._connect()
+
+        while True:
+            self._calculate.wait()
+            self._calculate.clear()
+            if self.isInterruptionRequested():
+                break
+            self.calculate_expression()
+        self._disconnect()
+
+    def callback_value(self, ch_index, value):
+        """
+        Callback executed when a channel receives a new value.
+
+        Parameters
+        ----------
+        ch_index : int
+            The channel index on the list for this rule.
+        value : any
+            The new value for this channel.
+
+        Returns
+        -------
+        None
+        """
+        self._values[ch_index] = value
+        if not self.connected:
+            self.warn_unconnected_channels()
+            return
+        self._calculate.set()
+
+
+    def callback_conn(self, ch_index, value):
+        """
+        Callback executed when a channel connection status is changed.
+
+        Parameters
+        ----------
+        ch_index : int
+            The channel index on the list for this rule.
+        value : bool
+            Whether or not this channel is connected.
+
+        """
+        self._connections[ch_index] = value
+        self._send_update(self.connected, self._value)
+
+    def warn_unconnected_channels(self):
```

Comment:
This is only called once - you could probably move this back into that method.

Reply:
Addressed via https://github.com/slaclab/pydm/pull/637/commits/af070cfc0f15fe9555b5bfa96b04ca6816de21ac